src/Arcus.EventGrid/Contracts/EventGridEvent.cs

```diff
@@ -0,0 +1,109 @@
+﻿using Arcus.EventGrid.Contracts.Interfaces;
+using GuardNet;
+using Microsoft.Azure.EventGrid.Models;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using System;
+
+namespace Arcus.EventGrid.Contracts
+{
+    /// <summary>
+    ///     Represents an event published on a topic.
+    ///     (Official schema documentation - https://docs.microsoft.com/en-us/azure/event-grid/event-schema#event-schema)
+    /// </summary>
+    /// <typeparam name="TData">Type of data payload</typeparam>
+    public class EventGridEvent<TData> : EventGridEvent, IEvent
```

Comment:
Or support both?

Reply:
I think the diff might be off here, but my suggestion would be to use it in the constructor when we are creating the object, otherwise it would be pointless indeed.

I've had a look at `Validate` and it throws exceptions which is fine for me personally, Gaurding does that as well.

What do you think?

Reply:
In that case we should assign the incoming values to the properties first and then call the validation method. Not sure where exactly this validation happens; but I guess it could be an extra way to make all the implementations to validate it via this method too.

Reply:
Maybe even make it virtual?

Reply:
> In that case we should assign the incoming values to the properties first and then call the validation method. 

The base constructor already assigns all the properties so we can just call the built-in `Validate` once our constructor is called, we don't really need our own validation imo.

https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.eventgrid.models.eventgridevent.validate?view=azure-dotnet#Microsoft_Azure_EventGrid_Models_EventGridEvent_Validate

> Not sure where exactly this validation happens; but I guess it could be an extra way to make all the implementations to validate it via this method too.

The validation occurs when sending if I recall correctly, but I'd suggest to already block it in the constructor.

Reply:
Ok, I now call the validation method in the ctor.

Reply:
That is correct, but we are overriding it and plugging in our own logic while we can just use the default logic.

Is there a reason why you want to define it yourself?

Reply:
Maybe with added properties, but than again; this could also be done in the ctor.
Nevermind.

Reply:
Agreed, let's use the built-in method and add the rest to ctor if we have to

Reply:
So not sure what to do now.
I remove the overridden validation method and place back the guards in the ctor?

Reply:
No, you can just delete the overriden validation method and use the built-in one which is called in ctor